### PR TITLE
Add Maximum Independent Set library

### PR DESCRIPTION
**Pull request by claud10cv**
Tarballs for MaximumIndependentSet (originally in C++)

### CODE REVIEW

File: M/MaximumIndependentSet/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+cd ${WORKSPACE}/srcdir/MaximumIndependentSet
+make -j${nproc}
+make install
+"""
+
+platforms = supported_platforms()
+filter!(t -> os(t) in ["linux"], platforms)
+filter!(t -> arch(t) == "x86_64", platforms)
+filter!(t -> isnothing(libc(t)) ||  libc(t) == "glibc", platforms)
+
```
Comment (imciner2) [inline M/MaximumIndependentSet/build_tarballs.jl:19]:
```suggestion
platforms = expand_cxxstring_abis(platforms)

```
The auditor reports that `std::string` is used, so we need to handle the ABI differences.

**Comment by imciner2** [comment]
The upstream makefile seems to have `-lcxxwrap_julia` on the link line for the dynamic library, I don't see that on the command line for the Yggdrasil build you did (and you also don't include the `libcxxwrap_julia_jll` as a dependency). Are you packaging the most recent version of the library, or purposefully holding it back?

File: M/MaximumIndependentSet/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+script = raw"""
+cd ${WORKSPACE}/srcdir/MaximumIndependentSet
+make -j${nproc}
+make install
+"""
+
+platforms = supported_platforms()
+filter!(t -> os(t) in ["linux"], platforms)
+filter!(t -> arch(t) == "x86_64", platforms)
+filter!(t -> isnothing(libc(t)) ||  libc(t) == "glibc", platforms)
```
Comment (imciner2) [inline M/MaximumIndependentSet/build_tarballs.jl:18]: This is a slightly awkward way of just specifying the one platform, but it works. Out of curiosity though, is there a reason to not build on the other platforms?

Comment (claud10cv) [reply M/MaximumIndependentSet/build_tarballs.jl:18]: The makefile is a hard-coded, hand-written one. It only works on linux for now. I will work on creating a cmake script to enable cross-compile later this week

File: M/MaximumIndependentSet/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+build_tarballs(ARGS,
+               name,
+               version,
+               sources,
+               script,
+               platforms,
+               products,
+               dependencies,
+               julia_compat = "1.9";
+               preferred_gcc_version=v"11")
```
[the remark this answers is not in the document]
Comment (claud10cv) [reply M/MaximumIndependentSet/build_tarballs.jl:36]: According to https://gcc.gnu.org/projects/cxx-status.html#cxx17 full support for c++17 is introduced as for g++-9. I will modify the line accordingly

File: M/MaximumIndependentSet/build_tarballs.jl (hunk 2 above)
Comment (claud10cv) [reply M/MaximumIndependentSet/build_tarballs.jl]: I am testing a cmake build and will see if I can get to coompile on more platforms

**Comment by claud10cv** [comment]
> The upstream makefile seems to have `-lcxxwrap_julia` on the link line for the dynamic library, I don't see that on the command line for the Yggdrasil build you did (and you also don't include the `libcxxwrap_julia_jll` as a dependency). Are you packaging the most recent version of the library, or purposefully holding it back?

Originally I had considered adding the wrapper within the C++ lib but ultimately decided to decouple both things, so I removed all references to libcxxwrap. I will create a separate wrapper package

File: M/MaximumIndependentSet/build_tarballs.jl (hunk 1 above)
Comment (claud10cv) [reply M/MaximumIndependentSet/build_tarballs.jl:24]: The tarballs now build for all platforms

File: M/MaximumIndependentSet/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+cmake --build build --parallel ${nproc}
+cmake --install build
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms(; 
+	exclude=[Platform("i686", "linux"; libc = "musl")
+		]
+)
```
Comment (giordano) [inline M/MaximumIndependentSet/build_tarballs.jl:27]: https://buildkite.com/julialang/yggdrasil/builds/9118#018e62ae-0e18-4c32-bb7b-99ef8051ce83/663-878
```
┌ Warning: /cache/build/yggy-amdci7-15/julialang/yggdrasil/M/MaximumIndependentSet/build/x86_64-linux-gnu/clphCAVm/x86_64-linux-gnu-libgfortran5-cxx11/destdir/lib/libmis.so contains std::string values!  This causes incompatibilities across the GCC 4/5 version boundary.  To remedy this, you must build a tarball for both GCC 4 and GCC 5.  To do this, immediately after your `platforms` definition in your `build_tarballs.jl` file, add the line:
│ 
│     platforms = expand_cxxstring_abis(platforms)
└ @ BinaryBuilder.Auditor /cache/julia-buildkite-plugin/depots/e2fd9734-29d8-45cd-b0eb-59f7104f3131/packages/BinaryBuilder/L8tBh/src/auditor/compiler_abi.jl:247
```
```suggestion
platforms = supported_platforms(; exclude=p->libc(p)=="musl" && arch(p)=="i686")
platforms = expand_cxxstring_abis(platforms)
```

**Review by giordano — COMMENTED** [review]
If you link to libjulia you really must follow what other packages do here, see for example: https://github.com/JuliaPackaging/Yggdrasil/blob/5dd8bee7317d0c5d0f797ae2328a069890eda5c4/L/libsingular_julia/build_tarballs.jl, since julia itself breaks the ABI in each minor release.

Highlights are
https://github.com/JuliaPackaging/Yggdrasil/blob/5dd8bee7317d0c5d0f797ae2328a069890eda5c4/L/libsingular_julia/build_tarballs.jl#L39-L44
(without the exclusion of Windows, if it isn't relevant for your case)
https://github.com/JuliaPackaging/Yggdrasil/blob/5dd8bee7317d0c5d0f797ae2328a069890eda5c4/L/libsingular_julia/build_tarballs.jl#L53
https://github.com/JuliaPackaging/Yggdrasil/blob/5dd8bee7317d0c5d0f797ae2328a069890eda5c4/L/libsingular_julia/build_tarballs.jl#L56

File: M/MaximumIndependentSet/build_tarballs.jl (hunk 4 above)
Comment (claud10cv) [reply M/MaximumIndependentSet/build_tarballs.jl]: I see. I believe I implemented all the changes suggested (except for the last one that clashes with on of the other suggestions above)

**Comment by giordano** [comment]
I forgot you need also # See https://github.com/JuliaLang/Pkg.jl/issues/2942
# Once this Pkg issue is resolved, this must be removed
uuid = Base.UUID("a83860b7-747b-57cf-bf1f-3e79990d037f")
delete!(Pkg.Types.get_last_stdlibs(v"1.6.3"), uuid)

**Comment by claud10cv** [comment]
> I forgot you need also
> 
> https://github.com/JuliaPackaging/Yggdrasil/blob/5dd8bee7317d0c5d0f797ae2328a069890eda5c4/L/libsingular_julia/build_tarballs.jl#L6-L9

Yep indeed without that the compiler complains about that UUID. Thank you for pointing out to the solution (or more like a workaround apparently)

File: M/MaximumIndependentSet/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+# Collection of sources required to complete build
+sources = [
+    GitSource("https://github.com/claud10cv/MaximumIndependentSet.git", "8a4d5f24f818800eb41ac4c7e8a3f663b96500d3")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/MaximumIndependentSet/src/
+Julia_PREFIX=${prefix}
+cmake -B build -DCMAKE_INSTALL_PREFIX=${prefix} -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} -DCMAKE_BUILD_TYPE=Release -DJulia_PREFIX=${Julia_PREFIX} -DJlCxx_DIR=$prefix/lib/cmake/JlCxx
```
Comment (giordano) [inline M/MaximumIndependentSet/build_tarballs.jl:22]: For readability:
```suggestion
cmake -B build \
    -DCMAKE_INSTALL_PREFIX=${prefix} \
    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
    -DCMAKE_BUILD_TYPE=Release \
    -DJulia_PREFIX=${Julia_PREFIX} \
    -DJlCxx_DIR=$prefix/lib/cmake/JlCxx
```

File: M/MaximumIndependentSet/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+
+include("../../L/libjulia/common.jl") 
+platforms = vcat(libjulia_platforms.(julia_versions)...) 
+platforms = expand_cxxstring_abis(platforms)
+
+# The products that we will ensure are always built
+products = Product[
+	LibraryProduct(["libmis", "mislib"], :mis)
```
Comment (giordano) [inline M/MaximumIndependentSet/build_tarballs.jl:36]:
```suggestion
products = [
    LibraryProduct(["libmis", "mislib"], :mis)
```

**Pull request merged**